style: Make restyle hints handle child combinators better than restyling the entire subtree

heycam opened the pull request:
<!-- Please describe your changes on the following line: -->

This PR reworks `RestyleHint` so that it can store the depth of the tree whose elements we need to re-run selector matching on, when selectors with `>` child combinators are involved.  For example, with `div.x > span > em`, if we toggle the `x` attribute on a `div`, we'll re-run selector matching on grandchildren of the `div`, rather than (as we do currently) on every descendants of the `div`.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/16906)
<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/data.rs, components/style/restyle_hints.rs, components/style/traversal.rs, components/style/matching.rs
 * @emilio: components/style/data.rs, components/style/restyle_hints.rs, ports/geckolib/glue.rs, components/style/traversal.rs, components/style/matching.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify style code, but no tests are modified. Please consider adding a test!

heycam commented:
r? @emilio

emilio commented on components/style/restyle_hints.rs:
```diff
@@ -110,39 +114,215 @@ pub fn assert_restyle_hints_match() {
+    /// restyle work, and thus any `insert()` calls will have no effect.
+    #[inline]
+    pub fn is_maximum(&self) -> bool {
+        self.match_self == 0xff && self.match_later_siblings && self.replacements.is_all()
```
what about using `std::u8::MAX` instead of `0xff`?

heycam replied:
I decided to leave this as the numeric literal, as I think it's slightly clearer given we're using it as a bitfield.

emilio commented on components/style/restyle_hints.rs:
```diff
@@ -847,12 +1035,12 @@ impl DependencySet {
             }
 
-            // We can ignore the selector flags, since they would have already
+            // We can ignore the selector replacements, since they would have already
```
This comment is talking about the slow selector flags, so I don't think it should be changed.

heycam replied:
Yeah, overzelaous find and replace.

emilio commented on components/style/restyle_hints.rs:
```diff
@@ -25,34 +25,44 @@ use smallvec::SmallVec;
+    /// children, the third its grandchildren, and so on.  If the most significant
+    /// bit it set, it indicates that that selector matching must be re-run for
+    /// elements at that depth and all of their descendants.
+    match_self: u8,
```
Could we abstract this and make it a newtype (`MatchingDepth(u8)`?) that enforces this? I really don't want to think a lot about what `0x01` means if I could do something like: `MatchingDepth::just_self()` or `MatchingDepth::at_depth(x)`.

Also, I don't think the name of `match_self` is representative anymore :)

heycam replied:
Great idea about the newtype!

Regarding the name, WDYT about `match_under_self`?  (I want a name that I can adjust to have a similar field to represent depth to restyle under later siblings, e.g. `match_later_siblings` or `match_under_later_siblings`.)

emilio replied:
`match_under_self` sounds good to me :)

emilio commented on components/style/restyle_hints.rs:
```diff
@@ -651,6 +819,8 @@ impl DependencySet {
         let mut iter = selector.inner.complex.iter();
         let mut index = 0;
+        let mut child_count = 0;
+        let mut descendants = false;
```
I'd rename this to `saw_descendant_combinator`.

heycam replied:
Just used short names to avoid wrapping lines that used them.  Will rename.

emilio commented on components/style/restyle_hints.rs:
```diff
@@ -685,16 +854,35 @@ impl DependencySet {
+                        if !descendants => RestyleHint::descendants_at_depth(child_count),
+                    Some(Combinator::Child) => RestyleHint::descendants(),
+                    Some(Combinator::Descendant) => RestyleHint::descendants(),
+                    Some(Combinator::NextSibling) => RestyleHint::later_siblings(),
```
You can merge some branches here if you want.

emilio commented on components/style/restyle_hints.rs:
```diff
@@ -651,6 +819,8 @@ impl DependencySet {
         let mut combinator = None;
         let mut iter = selector.inner.complex.iter();
         let mut index = 0;
+        let mut child_count = 0;
```
And perhaps these one to `child_combinators_seen`?

emilio reviewed (commented):
Looks great! mostly minor nits and naming shenanigans :)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #16900) made this pull request unmergeable. Please resolve the merge conflicts.

emilio reviewed (approved):
r=me, thanks!

jdm commented:
```
   Compiling style_tests v0.0.1 (file:///home/travis/build/servo/servo/tests/unit/style)

error: unexpected close delimiter: `}`

  --> /home/travis/build/servo/servo/tests/unit/style/restyle_hints.rs:27:1

   |

27 | }

   | ^

error: Could not compile `style_tests`.
```

heycam committed:
style: Refactor RestyleHint to be a struct.

Later PRs will add additional data to it that is not so easy to
represent using bitflags.

heycam committed:
style: List the matched rules with trace-level logging is enabled.

heycam committed:
style: Use RestyleDamage to determine whether we must continue cascading style changes to children.

heycam committed:
style: Generate RestyleHints to selector match at specific descendant depths.

heycam committed:
style: Record in RestyleHints how far down the tree selector matching must be re-run.

heycam commented:
I'm including the https://bugzilla.mozilla.org/show_bug.cgi?id=1289868 changes in this PR, which have also been reviewed by @emilio over there.

heycam commented:
@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit ad6a448 has been approved by `emilio`

<!-- @bors-servo r=emilio ad6a4488ba13714f3b8d3b81ddd7b9dcde8f4d98 -->

heycam commented:
@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit 4819dd9 has been approved by `emilio`

<!-- @bors-servo r=emilio 4819dd97122706c71ffa65c10c688ab2f8599f02 -->

heycam commented:
@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit ef21da7 has been approved by `emilio`

<!-- @bors-servo r=emilio ef21da796595b1610c3944749441976f6851a4ae -->

bors-servo commented:
:hourglass: Testing commit ef21da796595b1610c3944749441976f6851a4ae with merge 7b9f831ae86e0ea2ee3eb716e3b7aa987e50e0dc...

heycam commented:
@bors-servo r-

Something's gone wrong when I rebased and exports to my Servo repo. :(

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/7372)

heycam committed:
style: Add an outparam to Gecko_CalcStyleDifference that returns whether any style data changed.

heycam commented:
@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit 3b769a0 has been approved by `emilio`

<!-- @bors-servo r=emilio 3b769a08f0d5fefab4543d7f4a2aefc5e59000c5 -->

bors-servo commented:
:hourglass: Testing commit 3b769a08f0d5fefab4543d7f4a2aefc5e59000c5 with merge c80808501cdf3eb721fb878a433cb8faf5137bbf...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/4146)

heycam commented:
@bors-servo retry

* #15583

bors-servo commented:
:hourglass: Testing commit 3b769a08f0d5fefab4543d7f4a2aefc5e59000c5 with merge 0ceebf23f7fd4b12a09e282d1485107284887a92...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/4148)

heycam commented:
@bors-servo retry

heycam commented:
@bors-servo r-

Actually I think that's a real test failure. :(

heycam commented:
I kind of suspect that Servo's damage calculation might be missing some properties, and that's making our decision to stop restyling invalid.  For now I'm going to just always return `StyleChange::Changed` from `compute_style_difference` in Servo, and file a followup issue.

heycam commented:
@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit 3b769a0 has been approved by `emilio`

<!-- @bors-servo r=emilio 3b769a08f0d5fefab4543d7f4a2aefc5e59000c5 -->

heycam commented:
@bors-servo p=1

bors-servo commented:
:hourglass: Testing commit 3b769a08f0d5fefab4543d7f4a2aefc5e59000c5 with merge 2a8c46c5a872d7f80a5883685b0b10d877b2874e...

emilio commented:
@bors-servo retry

bors-servo commented:
:hourglass: Testing commit 3b769a08f0d5fefab4543d7f4a2aefc5e59000c5 with merge ff76c4fb99b5afa264dd3f3cf9b2e2a4dfdac9c5...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/4154)

emilio commented:
Rebased at #16967.

@heycam, you should be able to push to that branch, if I haven't managed to land it, please feel free to take over it.

emilio commented:
I posted the analysis of the failing test there, too, btw

emilio commented:
Landed #16967 :)

heycam commented:
Thanks so much, @emilio!
